### PR TITLE
T1790

### PR DESCRIPTION
**Pull request by k88hudson**

### CODE REVIEW

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @k88hudson received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/d2327334a0bc0ad/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/d2327334a0bc0ad/output.txt

Total script time: 1.19 mins

Server is running on http://pr793.butyr.org/templates/basic/
Run the Test Suite: http://pr793.butyr.org/test/

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @k88hudson received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/46dfd49ce38c9f5/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/46dfd49ce38c9f5/output.txt

Total script time: 1.07 mins

Server is running on http://pr793.butyr.org/templates/basic/
Run the Test Suite: http://pr793.butyr.org/test/

File: src/layouts/ui-kit.html [hunk 1]
```diff
@@ -31,6 +32,26 @@
...
+var tooltip1 = Tooltip.create({ 
+  message: "hello world" 
+});
+myButton.appendChild( tooltip1 );
+
+// Apply to all elements inside a rootElement  
+&lt;a id="test2" class="butter-btn btn-light" 
+  data-tooltip="coolio"&gt;Button&lt;/a&gt;
+  
+Tooltip.apply( rootElement );
```
Comment (dseif) [inline src/layouts/ui-kit.html:49]: indentation is off for this block

File: src/ui/widget/tooltip.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
+/* This Source Code Form is subject to the terms of the MIT license
+ * If a copy of the MIT license was not distributed with this file, you can
+ * obtain one at http://www.mozillapopcorn.org/butter-license.txt */
+
+define( [], function() {
```
Comment (dseif) [inline src/ui/widget/tooltip.js]: Mind adding a comment here to explain what this file is used for

File: src/ui/widget/tooltip.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+  var __tooltipClass = "butter-tooltip",
+      __tooltipOnClass = "tooltip-on";
+
+  var Tooltip = {
```
Comment (dseif) [inline src/ui/widget/tooltip.js]: these two var blocks can become one

File: src/ui/widget/tooltip.js [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+      tooltipEl.classList.add( __tooltipClass );
+      tooltipEl.innerHTML = tooltipText;
+
+      if ( options.hidden === false ) {
+        tooltipEl.classList.add( __tooltipOnClass );
+      }
+
+      if ( element ) {
+         // Parent must be relative or absolute for tooltip to be positioned properly
+        if ( getComputedStyle( element ).getPropertyValue( "position" ) !== "absolute" || "relative" ) {
```
Comment (dseif) [inline src/ui/widget/tooltip.js]: I don't think this actually works.  Try doing:
if ( `[ '"absolute", "relative" ].indexOf( getComputedStyle( element ).getPropertyValue( "position" ) ) == -1 )`

File: src/ui/widget/tooltip.js [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+      return tooltipEl;
+    },
+
+    /**
+     * Member: apply
+     *
+     * Creates a tooltip inside all elements of a given root element with data-tooltip attribute
+     */
+    apply: function( rootElement ) {
+      var elements, i, l;
```
Comment (dseif) [inline src/ui/widget/tooltip.js]: make this multi-lined please

File: src/ui/widget/tooltip.js [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+      rootElement = rootElement || document;
+      elements = rootElement.querySelectorAll( "[data-tooltip]" );
+
+      for ( i = 0, l = elements.length; i < l; i++ ) {
+        Tooltip.create({
+          element: elements[ i ]
+        });
+      }
+    }
+
```
Comment (dseif) [inline src/ui/widget/tooltip.js]: remove whitespace

File: src/ui/widget/tooltip.js [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+  };
+  
+  return Tooltip;
+
```
Comment (dseif) [inline src/ui/widget/tooltip.js]: remove whitespace

File: src/ui/widget/tooltip.js [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+        parentRect = element.getBoundingClientRect();
+        tooltipEl.style.top = top || parentRect.height + "px";
+        tooltipEl.style.left = left || "50%";
+
+        element.appendChild( tooltipEl );
+      }
+
+      return tooltipEl;
+    },
+
```
Comment (dseif) [inline src/ui/widget/tooltip.js]: remove whitespace

File: src/editor/base-editor.js [hunk 9]
```diff
@@ -127,6 +130,15 @@ define( [ "core/eventmanager", "util/scrollbars" ],
       return extendObject.scrollBar;
     };
 
+     /**
```
Comment (dseif) [inline src/editor/base-editor.js]: This line is indented 1 space too far

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_check** from @dseif received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/9a99a9b78efed2f/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/9a99a9b78efed2f/output.txt

Total script time: 1.00 mins

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @dseif received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/abdc9a8528a626a/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/abdc9a8528a626a/output.txt

Total script time: 0.76 mins

Server is running on http://pr793.butyr.org/templates/basic/
Run the Test Suite: http://pr793.butyr.org/test/

File: src/layouts/ui-kit.html (hunk 1 above)
Comment (k88hudson) [inline src/layouts/ui-kit.html:49]: I have to because it's in a `<pre>` tag

Comment (dseif) [inline src/layouts/ui-kit.html:49]: hahaha my bad, I'm retarded. Carry on.

File: src/layouts/ui-kit.html [hunk 10]
```diff
@@ -8,11 +8,12 @@
       <div class="editor-options scrollbar-inner">
         <fieldset>
           <label>Buttons</label>
-          <a class="butter-btn btn-light">Button</a>
-          <a class="butter-btn btn-green">Button</a>
+          <a class="butter-btn btn-light" data-tooltip="This is an example of a tooltip">Button
+          </a>
```
Comment (dseif) [inline src/layouts/ui-kit.html]: Move this up to the line above

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_check** from @cadecairos received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/70d4451a6dd24aa/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @cadecairos received. Current queue size: 1

Live output at: http://botio.butyr.org:8000/65f6086b55d67bf/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @cadecairos received. Current queue size: 2

Live output at: http://botio.butyr.org:8000/82482650bc15119/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Failed

Full output at http://botio.butyr.org:8000/70d4451a6dd24aa/output.txt

Total script time: 1.02 mins

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/65f6086b55d67bf/output.txt

Total script time: 0.59 mins

Server is running on http://pr793.butyr.org/templates/basic/
Run the Test Suite: http://pr793.butyr.org/test/

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/82482650bc15119/output.txt

Total script time: 0.68 mins

Server is running on http://pr793.butyr.org/templates/basic/
Run the Test Suite: http://pr793.butyr.org/test/

File: external/popcorn-js [hunk 11]
```diff
@@ -1 +1 @@
-Subproject commit 7bc08e1641eacd76e3e3ddf957500a805696504a
+Subproject commit f4b273ca9eb0a5c0a060c332a3e19cddc5d9fdc3
```
Comment (jbuck) [inline external/popcorn-js]: undo this

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @jbuck received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/84ca5ef99d6f641/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/84ca5ef99d6f641/output.txt

Total script time: 0.67 mins

Server is running on http://pr793.butyr.org/templates/basic/
Run the Test Suite: http://pr793.butyr.org/test/

**Pull request merged**